### PR TITLE
ARROW-8142: [C++][Compute] Explicit no chunks case for WrapDatumsLike

### PR DESCRIPTION
**Pull request by bkietz**
When kernels which use the `WrapDatumsLike` utility on a chunked array, they invoke the ChunkedArray constructor without an explicit type. This [is forbidden for an empty chunks vector](https://github.com/bkietz/arrow/blob/1cb7a6dcbfc61a734e0e888917fa3938051de82d/cpp/src/arrow/table.h#L41) and causes an assertion failure. Since we always have access to the desired type (through Kernel::out_type), we can avoid this by passing that type to WrapDatumsLike.

### CODE REVIEW

**Comment by jorisvandenbossche** [comment]
The aborting tests seem to be limited to several tests dealing with string to dictionary conversion in `to_pandas`:

python/pyarrow/tests/test_pandas.py::TestConvertStringLikeTypes::test_table_empty_str 
python/pyarrow/tests/test_pandas.py::TestConvertStringLikeTypes::test_selective_categoricals 
python/pyarrow/tests/test_pandas.py::TestConvertStringLikeTypes::test_table_str_to_categorical_without_na 
 
Smaller reproducer:

```
In [1]: table = pa.table([pa.array(['a', 'b'], type=pa.string())], ['col1']) 

In [2]: table.to_pandas(strings_to_categorical=True)     
terminate called after throwing an instance of 'std::bad_cast'
  what():  std::bad_cast
Aborted (core dumped)
```

In addition, there is also one failing test (python/pyarrow/tests/test_parquet.py::test_direct_read_dictionary), which seems to be due to the table constructor loosing the type of a dictionary encoded chunked array:

```
In [12]: chunked_array = pa.chunked_array([pa.array(["a", "b"])]).dictionary_encode()  

In [13]: pa.table({'f0': chunked_array})   
Out[13]: 
pyarrow.Table
f0: int32
```

**Comment by jorisvandenbossche** [comment]
The second failure comes from this (which might help finding the root cause):

```
In [8]: chunked_array = pa.chunked_array([pa.array(["a", "b"])]).dictionary_encode() 

In [9]: chunked_array   
Out[9]: 
<pyarrow.lib.ChunkedArray object at 0x7f19c6c99b88>
[

  -- dictionary:
    [
      "a",
      "b"
    ]
  -- indices:
    [
      0,
      1
    ]
]

In [10]: chunked_array.type   
Out[10]: DataType(int32)

In [11]: chunked_array.chunk(0).type
Out[11]: DictionaryType(dictionary<values=string, indices=int32, ordered=0>)
``

So when dicitionary encoding a chunked array, the type of the chunked array is set incorrectly (the type of the individual array chunks is correct)

So I suppose that the `out_type()` passed in cast.cc::DictionaryEncode is the wrong type.

File: cpp/src/arrow/compute/kernels/hash.cc [hunk 1]
```diff
@@ -591,7 +591,7 @@ Status DictionaryEncode(FunctionContext* ctx, const Datum& value, Datum* out) {
       dict_chunks.emplace_back(std::make_shared<DictionaryArray>(
           dict_type, MakeArray(datum.array()), dictionary));
     }
-    *out = detail::WrapArraysLike(value, dict_chunks);
+    *out = detail::WrapArraysLike(value, func->out_type(), dict_chunks);
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/compute/kernels/hash.cc]: I suppose this `out_type` is incorrect (it seems to be the type of the indices)

Comment (bkietz) [reply cpp/src/arrow/compute/kernels/hash.cc]: @jorisvandenbossche you're correct. Also the hash kernel had a conflicting fix for the same basic problem from #5650 which should be resolved now

Comment (bkietz) [reply cpp/src/arrow/compute/kernels/hash.cc]: Thanks for your thorough investigation

**Comment by jorisvandenbossche** [comment]
@bkietz added two python tests as well (which cover the reproducers from the two JIRA issues)

**Comment by bkietz** [comment]
@jorisvandenbossche thanks!

**Review by pitrou — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**